### PR TITLE
WIP : Celer numba

### PR DESCRIPTION
**Pull request by mathurinm**

### CODE REVIEW

File: solvers/celer_numba.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+            # compute the eigenvalue and stop if it does not move anymore
+            rk1 = rk
+            rk = b_k1 @ b_k
+            if abs(rk - rk1) < 1e-10:
+                break
+
+            # re normalize the vector
+            b_k = b_k1 / np.linalg.norm(b_k1)
+
+        return rk
```
Comment (agramfort) [inline solvers/celer_numba.py:313]:
```suggestion
    def compute_lipschitz_constant(self):
        if not sparse.issparse(self.X):
            L = np.linalg.norm(self.X, ord=2) ** 2
        else:
            L = sparse.linalg.svds(self.X, k=1)[1][0] ** 2
        return L
```

File: solvers/celer_numba.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    def set_objective(self, X, y, lmbd):
+        self.X, self.y, self.lmbd = X, y, lmbd
+
+    def run(self, n_iter):
+        w = numba_celer(self.X, self.y, self.alpha, n_iter)
+        self.w = w
+
+    def st(self, w, mu):
+        w -= np.clip(w, -mu, mu)
+        return w
```
Comment (agramfort) [inline solvers/celer_numba.py:281]: use ST above?

File: solvers/celer_numba.py (hunk 1 above)
Comment (mathurinm) [reply solvers/celer_numba.py:313]: Thanks, but there's a PR in master doing that, this part will be deleted anyway

**Comment by mathurinm** [comment]
Ok it seems to give the exact same results as celer main branch:


python
```import numpy as np
from numpy.linalg import norm

from celer.datasets import make_correlated_data
from celer import Lasso

X, y, _ = make_correlated_data(n_samples=200, n_features=1000, random_state=0)

y /= norm(y) / np.sqrt(len(y))

alpha = np.max(np.abs(X.T @ y)) / len(y) / 20

max_iter = 2
res = numba_celer(X, y, alpha, n_iter=max_iter, verbose=2, tol=1e-10)


clf = Lasso(alpha=alpha, fit_intercept=False,
            verbose=2, max_iter=max_iter, tol=1e-10).fit(X, y)


np.testing.assert_allclose(res, clf.coef_)
```


Output:
python
```
Iter 0: primal 0.5000000000, gap 4.51e-01, 10 feats in subpb (1000 left)
Epoch 10, primal 0.3463196589, gap: 3.38e-05
Exit epoch 10, gap: 3.38e-05 < 1.35e-01
Iter 1: primal 0.3463196589, gap 2.89e-01, 20 feats in subpb (1000 left)
Epoch 10, primal 0.2778314574, gap: 7.65e-06
Exit epoch 10, gap: 7.65e-06 < 8.68e-02
#########################
##### Computing alpha 1/1
#########################
Iter 0: primal 0.5000000000, gap 4.51e-01, 10 feats in subpb (1000 left)
Epoch 10, primal 0.3463196589, gap: 3.38e-05
Exit epoch 10, gap: 3.38e-05 < 1.35e-01
Iter 1: primal 0.3463196589, gap 2.89e-01, 20 feats in subpb (1000 left)
Epoch 10, primal 0.2778314574, gap: 7.65e-06
Exit epoch 10, gap: 7.65e-06 < 8.68e-02
```

**Comment by agramfort** [comment]
can you share a comparison in perf between cython and numba?

**Comment by PABannier** [comment]
@mathurinm indeed, I did a mess with my linter. I disabled it to avoid large diffs.

**Comment by PABannier** [comment]
@mathurinm yeah I can create a small package to avoid duplicate code

**Comment by mathurinm** [comment]
I merge now since it works, and we can merge the primal and dual solvers in a subsequent PR.

**Pull request merged**